sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java

```diff
@@ -753,17 +757,44 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
                     result = result.contextWrite(context -> context.put(ENCRYPTION_DATA_KEY,
                         EncryptionData.getAndValidateEncryptionData(encryptionDataKey, requiresEncryption)));
                 }
+                if (isDownloadToFile) {
+                    result = result.contextWrite(context -> context.put("rangeAdjustment", adjustRangeThroughContext()));
+                }
                 return result;
             });
     }
 
+    private BiFunction<BlobDownloadAsyncResponse, Long, Long> adjustRangeThroughContext() {
+        return (response, totalLength) -> {
+            long newLength = totalLength;
+            String metadetaString = response.getDeserializedHeaders().getMetadata().get("encryptiondata");
+            if (metadetaString == null) {
+                return newLength;
+            }
+            EncryptionData encryptionData;
+            try (JsonReader jsonReader = JsonProviders.createReader(metadetaString)) {
+                encryptionData = EncryptionData.fromJson(jsonReader);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            //if the blob is encrypted using V2, use the unencrypted length for range calculations
+            if ("2.0".equals(encryptionData.getEncryptionAgent().getProtocol())) {
+                long regionLength = getClientSideEncryptionOptions().getAuthenticatedRegionDataLengthInBytes();
+                long region = Math.floorDiv(totalLength, regionLength);
```

Comment:
For this calculation, we should double check if we need to be doing floor or ceiling division. I'd imagine it would be ceiling as in the case where the total length fits in one encryption region, the total unencrypted length should be just the `total length - (nonce + tag)` and here it seems like it would just return the encrypted length?

Once we break this out to a separate class, we should be able to parameterize around these edge case in a unit test for the class to make sure we are confident it is behaving correctly.